### ISSUE TITLE
Composer.lock

### ISSUE THREAD

**Issue by AwesomeBobX64** [google/google-api-php-client#991]
You locked the thread for the PR this issue is based off of so I'm filing an issue (https://github.com/google/google-api-php-client/pull/973).

https://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file

"After installing the dependencies, Composer writes the list of the exact versions it installed into a composer.lock file. This locks the project to those specific versions.

**Commit your application's composer.lock (along with composer.json) into version control.**

This is important because the install command checks if a lock file is present, and if it is, it downloads the versions specified there (regardless of what composer.json says).

This means that anyone who sets up the project will download the exact same version of the dependencies. Your CI server, production machines, other developers in your team, everything and everyone runs on the same dependencies, which mitigates the potential for bugs affecting only some parts of the deployments. Even if you develop alone, in six months when reinstalling the project you can feel confident the dependencies installed are still working even if your dependencies released many new versions since then."

The Composer documentation highlights why this is imporant and calls out the issue we're experiencing from your lack of understanding. You obviously haven't read the Composer documentation.

**Comment by tmatsuo** [comment]
https://getcomposer.org/doc/02-libraries.md#lock-file

For libraries, it makes sense to not have composer.lock file in the repo, because the composer.lock file in libraries won't lock your application's dependencies. In my opinion, libraries should have as loose restriction as possible to allow maximum flexibility on application that rely on them.

**Comment by AwesomeBobX64** [comment]
Okay then, rather than committing the composer.lock file can you fix the linking of the dependencies in the composer.json and to follow the semantic versioning specifications so that bugs aren't introduced when a dependency increases in a minor version?

Just installing v2.0.0 of the google/apiclient package on two different days of the same week resulted in breaking changes due to the loose coupling of different versions of the dependencies.

Binding to @dev is too lose and allowed breaking changes between v0.2 and v0.3 to make the library unusable. It should not be the responsibility of every consumer of the package to handle the introduction of breaking changes when installing the same version of the package.

**Comment by tmatsuo** [comment]
Ideally, each api service package should be an independent composer package because API release should conform to semantic versioning.

Even if that happens, it is still a responsibility of application to stick to a specific version.

For example, if you use php-datastore-client v2.0.0 and the datastore releases v3.0.0 API, we will start having it as a default. This is because we want to provide the latest version by default.

Unfortunately, currently `google/apiclient-services` contains all of the services. Potentially we can just tag it as 1.0.0, and weekly increment the major version (potentially we can check the API version change, but it's a bit of work).

Then still, application has a responsibility of sticking to a specific release tag of `google/apiclient-services`. Normally, if you don't want to be annoyed by any compatibility changes, your best bet is to check the current latest version on the [repo,](https://github.com/google/google-api-php-client-services/releases) and use that specific version of `google/apiclient-services`.

I hope it does alleviate your issue.

Re: semantic versioning
Basically what you're saying is right. Ideally we should follow this, but because of the current way of how `google/apiclient-services` package is built, it's a bit hard to always provide the most appropriate version number. So I'm proposing weekly major version increases.


Last but not least, @bshaffer did a great job of automating the release of `google/apiclient-services`. I would like to ask you to be less oppressive. The current state is not ideal, but it's significantly improved than it was before.

If we have indefinite time, we can provide ideal situation, but unfortunately our resource is limited and the current form is a result of some pragmatic compromises.

I hope this helps you to understand the current situation.

**Comment by bshaffer** [comment]
@AwesomeBobX64 We appreciate your feedback on how tying `google/apiclient-services` to `@dev` has been inconvenient for you.  I've added https://github.com/google/google-api-php-client-services/issues/23 to track a feature that will help remove `@dev` while still allowing users to keep their APIs up-to-date.

As stated, `composer.lock` would have no affect on projects that depend on this package, so I am closing this issue.
